daemon/task.go

```diff
@@ -103,44 +112,102 @@ func FormatTime(t time.Time) string {
 
 // Map the task onto a map[string]interface{}, using the given route for the Location()
 func (t *Task) Map(route *mux.Route) map[string]interface{} {
+	t.Lock()
+	defer t.Unlock()
+
 	return map[string]interface{}{
 		"resource":   t.Location(route),
 		"status":     t.State(),
-		"created_at": FormatTime(t.CreatedAt()),
-		"updated_at": FormatTime(t.UpdatedAt()),
+		"created_at": FormatTime(t.t0),
+		"updated_at": FormatTime(t.tf),
 		"may_cancel": false,
-		"output":     t.Output(),
+		"output":     t.output,
+	}
+}
+
+var (
+	// ErrNoReceiver is returned from Send when the task isn't listening (yet)
+	ErrNoReceiver = errors.New("task is not receiving messages at this time")
+	// ErrNotRunning is returned from Send when the task isn't Running
+	ErrNotRunning = errors.New("task is not running")
+)
+
+// Send a value into the task
+func (t *Task) Send(v interface{}) error {
+	t.Lock()
+	defer t.Unlock()
+
+	if t.State() != TaskRunning {
+		return ErrNotRunning
+	}
+
+	if t.chin == nil {
```

Comment:
I would be surprised to get an ErrNoReceiver error just because the goroutine did not run fast enough to reach the point where a channel is assigned. This is the sort of race we should avoid in the name of easier concurrency.

To solve it, I suggest changing RunTask to take the function as an interface{}, and then casting the value to one of the three following types:

    func() interface{}             // Old school version
    func(out chan interface{})     // Can produce multiple outputs
    func(in, out chan interface{}) // Can take inputs and produce multiple outputs

This enables you to create the channels while the task is still locked down and before the goroutine is spawned. The distinction between the second and third option also prevents the issue that will happen when you want to have a task that can output multiple values but is not interested in taking inputs, so that we can error out early instead of being blocked for the whole duration of the task when we know the task is not going to consume the input.